syz-cluster: improve series triage

a-nogikh opened the pull request:
* Be more explicit about the exact reason for skipping a series.
* Ignore the series that do not modify any `.h` or `.c` files.

tarasmadan commented on syz-cluster/pkg/triage/commit.go:
```diff
@@ -21,26 +21,37 @@ type TreeOps interface {
-		log.Printf("the series is too old: %v before the HEAD", diff)
-		return "", nil
+		cs.tracer.Log("the series is too old: %v before the HEAD", diff)
+		return SelectResult{Reason: reasonSeriesTooOld}, nil
```
It looks confusing.
`reasonSeriesTooOld` looks like an explanation of why you can't return `Commit`.
`return "", reasonSeriesTooOld` may be a better way.

tarasmadan replied:
Agree. And the final decision is up to you. You can get 1 more opinion from LLMs :)

tarasmadan commented on syz-cluster/pkg/triage/diff.go:
```diff
@@ -0,0 +1,34 @@
+// 2. Some source files are compiled but we have never been able to cover them during fuzzing.
+// 3. Some non-.c and non-.h files are used to auto-generated the source code.
+
+const reasonNotAffectsBuild = "the series doesn't modify .c and .h files"
```
Makefile, Kconfig, S and probably some others may also be the candidates.

```bash
torvalds$ find . -type f | awk -F '[./]+' '{print $NF}' | sort | uniq -c | sort -rn
  34761 c
  25348 h
   4436 yaml
   3629 rst
   3136 dts
   3036 Makefile
   2362 dtsi
   1732 Kconfig
   1504 txt
   1334 S
    913 sh
    867 json
    359 gitignore
    264 py
    178 config
    151 dtso
    133 Kbuild
    127 j2
    126 tc
     86 Build
     82 rs
     79 svg
     73 cocci
     59 README
     55 pl
     47 litmus
     41 TODO
     35 boot
     31 settings
     30 pkt
     30 debug
     30 bconf
     28 Platform
     24 xml
     23 tbl
     23 log
     22 dot
     21 kunitconfig
     21 conf
     15 sample
     15 cfg
     15 asn1
     14 fuc
     13 1
     12 fuc3
     11 ppm
     11 csv
     11 awk
     10 y
     10 xsl
     10 l
     10 8
      9 syscalls
      8 map
      8 lds
      8 exceptions
      8 asm
      7 sa
      7 po
      7 include
      7 h_shipped

```

dvyukov replied:
> Makefile, Kconfig, S and probably some others may also be the candidates.

+1

Also .rs, scripts that do codegen, etc.

On the other hand, there are *.c files in tests, docs, perf, and lots of changes in kernel .c files that don't affect builds (comments, ifdefs, whitespace, trivial changes).

If we go the route of second guessing if a particular patch affects build or not, I afraid it's beginning of slippery slope.
We already need to solve this problem for bisection, and we use kernel binary signatures:
https://github.com/google/syzkaller/blob/1ef3ab4dee83617d4f9652edc2ddf2e0e8988c9a/pkg/build/linux.go#L252-L285
They handle all of the above positive and negative cases, even including some effectively on-op code changes.

tarasmadan replied:
How many patch series do you expect to exclude? Do you have %?

a-nogikh replied:
I don't have the %.

Moved the commit to #5934.

dvyukov commented:
> Elf binary signature is reasonable indeed, though the question is whether it's 100% reliable now - https://github.com/google/syzkaller/issues/2297

Let's say it's not, what does it mean in this context?

> One more way would be to consider the .d files created by make

I would strongly object to that. That's a more complex solution to a problem for which we already have a simpler and better solution. It depends on build systems that use make. It will produce false positives on comment patches, ifdef'ed patches, patches that do cosmetic changes to source code, etc.

a-nogikh commented:
I've changed the logic to "skip the series if it only modifies the files of the specified extensions".

And I've updated the description of the `syz-cluster: share the series skip reason` commit.
